lib/router/src/cache.rs

```diff
@@ -1,426 +1,367 @@
         assert!(cache.access(&1).is_some());
         assert!(cache.access(&2).is_some());
+
+        assert_eq!(cache.access(&1).take().unwrap().node.value, 2);
+        assert_eq!(cache.access(&2).take().unwrap().node.value, 3);
     }
 
     #[test]
     fn reserve_does_nothing_when_capacity_exists() {
-        let mut cache = Cache::<_, MultiplyAndAssign, _>::new(2, Duration::from_secs(0));
+        let (cache, _cache_purge) = Cache::new(2, Duration::from_millis(10));
 
-        // Create a route that goes idle immediately:
-        {
-            let r = cache.reserve().expect("capacity");
-            let mut service = MultiplyAndAssign::default();
-            service.call(1.into()).wait().unwrap();
-            r.store(1, service);
-        };
+        let mut cache = cache.lock().unwrap();
+
+        cache.reserve().expect("capacity").store(1, 2);
         assert_eq!(cache.vals.len(), 1);
 
         assert!(cache.reserve().is_ok());
         assert_eq!(cache.vals.len(), 1);
     }
 
     #[test]
-    fn reserve_honors_max_idle_age() {
-        let mut clock = Clock::default();
-        let mut cache = Cache::<_, MultiplyAndAssign, _>::new(1, Duration::from_secs(2))
-            .with_clock(clock.clone());
-
-        // Touch `1` at 0s.
-        cache
-            .reserve()
-            .expect("capacity")
-            .store(1, MultiplyAndAssign::default());
-        assert_eq!(
-            cache.reserve().err(),
-            Some(CapacityExhausted { capacity: 1 })
-        );
-        assert_eq!(cache.vals.len(), 1);
-
-        // No capacity at 1s.
-        clock.advance(Duration::from_secs(1));
-        assert_eq!(
-            cache.reserve().err(),
-            Some(CapacityExhausted { capacity: 1 })
-        );
-        assert_eq!(cache.vals.len(), 1);
-
-        // No capacity at 2s.
-        clock.advance(Duration::from_secs(1));
-        assert_eq!(
-            cache.reserve().err(),
-            Some(CapacityExhausted { capacity: 1 })
-        );
-        assert_eq!(cache.vals.len(), 1);
+    fn store_and_self_purge() {
+        tokio::run(futures::lazy(|| {
+            let (cache, _cache_purge) = Cache::new(2, Duration::from_millis(10));
+
+            // Do not spawn background purge. Instead, fill the cache and
+            // force a manual purge to occur when a value can be expired.
+            {
+                let mut cache = cache.lock().unwrap();
+
+                cache.reserve().expect("reserve").store(1, 2);
+                cache.reserve().expect("reserve").store(2, 3);
+                assert_eq!(cache.vals.len(), 2);
+            }
 
-        // Capacity at 3+s.
-        clock.advance(Duration::from_secs(1));
-        assert!(cache.reserve().is_ok());
-        assert_eq!(cache.vals.len(), 0);
+            // Sleep for an amount of time greater than expiry duration so
+            // that a value can be purged.
+            tokio::spawn(
+                tokio_timer::sleep(Duration::from_millis(100))
```

Comment:
Now that the [`tokio-test` crate](https://github.com/tokio-rs/tokio/tree/master/tokio-test) has been released, we might want to look into rewriting this (and other tests) to use mock timers instead. It would be fine to do this in a follow up branch though.

Reply:
I'd like to include this in linkerd/linkerd2#2802